Collect I/O metrics from disk partitions and loop devices

gdemonet opened the pull request:
**Component**: node-exporter, tests

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: See #2631 (we need metrics for all kinds of devices that can be used in Volumes)

**Summary**:

- Adapt `node-exporter`'s `collector.diskstats.ignored-devices` to only ignore RAM disks and floppy disks
- Add a test to ensure we can actually get metrics for a `SparseLoopDevice` Volume

**Acceptance criteria**: 

- tests pass
- metrics can be retrieved for all kinds of devices allowed in Volumes

_Note that those metrics are always 0 for loop devices on old kernels such as what comes with CentOS 7_

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2631 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

gdemonet committed:
charts: Do not ignore partitions or loop devices

We override the default list of ignored devices for `node-exporter`'s
diskstats collector.
This will allow our UI to provide useful stats at the Volume level.

Fixes: #2631

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2631-collect-metrics-from-partitions` into
`development/2.5`
* [w/2.6/improvement/2631-collect-metrics-from-partitions](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/2631-collect-metrics-from-partitions) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

gdemonet commented on tests/post/steps/test_volume.py:
```diff
@@ -93,23 +91,27 @@
 @pytest.fixture
 def pv_client(k8s_client):
-    return PersistentVolumeClient(k8s_client)
+    return kube_utils.Persistentkube_utils.VolumeClient(k8s_client)
```
that can't work :facepalm:

bert-e commented:
# History mismatch



Merge commit #0c5f7cfd7f6de9ab81267a32a61f6eb775f1f54a on the integration branch
`w/2.6/improvement/2631-collect-metrics-from-partitions` is merging a branch which is neither the current
branch `improvement/2631-collect-metrics-from-partitions` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/2631-collect-metrics-from-partitions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented on tests/post/features/monitoring.feature:
```diff
@@ -42,4 +42,10 @@ Feature: Monitoring is up and running
\ No newline at end of file
+        Then the deployed Prometheus alert rules are the same as the default alert rules
+
+    Scenario: Volume metrics can be found based on device name
+        Given the Kubernetes API is available
+        And I have created a test Volume
+        And the Prometheus API is available
+        Then I can get I/O stats for the test Volume device
```
To make the test faster/less flaky, I guess you could skip the creation of a brand new volume and try to get the stats from one of the two already existing volume (alertmanager's one or prometheus' one).

gdemonet replied:
Right... But I didn't want to rely on the fact that this volume _may_ be using a loop device (ideally, our `post` E2E tests are non intrusive, so we can run them to perform a sanity check against a "real" cluster).

slaperche-scality replied:
ha yeah right, we want to check against a loop device specifically (since attaching a real partitioned volume would be even flakier…)

slaperche-scality reviewed (commented):
minor remark: I would have done the code moving/refactoring in a dedicated commit, to have a smaller commit introducing the new tests.

gdemonet commented:
> I would have done the code moving/refactoring in a dedicated commit

Good point, will do that!

gdemonet committed:
tests: Use control_plane_ip fixture when relevant

We already had a fixture ready but were still fetching the grain once
again in the `prometheus_api` fixture.

gdemonet committed:
tests: Extract utility clients for K8s API

These were only defined in `test_volume`, which makes it harder for
other modules to re-use (even though they are quite handy!).

gdemonet committed:
tests: Support Prometheus query in utility client

The previous implementation called `query` a plain "request" to
Prometheus API. This gets renamed `request` (surprising!), and we add a
`query` method for performing a call to /api/v1/query.

See
https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries
for details.

gdemonet committed:
tests: Check that we have I/O stats for Volumes

This test creates a test Volume using sparse loop device, and ensures
the `status.deviceName` can be used to retrieve read/write metrics from
Prometheus.

See: #2631

slaperche-scality commented on tests/post/steps/test_monitoring.py:
```diff
@@ -81,6 +96,35 @@ def _check_object_exists():
 
 
+# FIXME: make these fixtures / helpers generic and easier to share
+@given("I have created a test Volume")
+def test_volume(k8s_apiclient, ssh_config):
+    client = kube_utils.VolumeClient(
+        CustomObjectsApi(api_client=k8s_apiclient), ssh_config
+    )
+    random_salt = ''.join(
+        random.choice(string.ascii_lowercase) for _ in range(8)
+    )
+    volume_name = "test-volume-{}".format(random_salt)
+    client.create_from_yaml(DEFAULT_VOLUME.format(name=volume_name))
+
+    def _check_volume_ready():
+        volume = client.get(volume_name)
+        assert volume is not None, 'Volume not found'
+        assert 'status' in volume, 'Volume has no status'
+        phase = kube_utils.VolumeClient.compute_phase(volume['status'])
+        assert phase == 'Available', 'Volume not ready'
+        assert volume['status'].get('deviceName') is not None, \
+            'Volume status.deviceName has not been reconciled'
+        return volume
+
+    yield utils.retry(
+        _check_volume_ready, times=30, wait=2,
+        name='waiting for Volume {} to be ready'.format(volume_name)
+    )
+
+    client.delete(volume_name, sync=True)
```
Can't we reuse the existing `@given(parsers.parse("a Volume '{name}' exist"))`
Looks very similar.

If not the step directly, at least the implementation (to reduce code duplication)

gdemonet replied:
Can do!

gdemonet replied:
Did something, but couldn't factorize everything (the test Volume doesn't have a static name, and the teardown is handled directly in the `given` fixture).

slaperche-scality replied:
Do we really need the random name?
We could use a simple naming scheme, like `test-monitoring-volumeN`?

gdemonet replied:
Most of the logic was factorized in the `VolumeClient` directly :slightly_smiling_face: but still not sharing the fixture/step because of how Volume tests rely on the `context` fixture.

slaperche-scality commented on tests/post/steps/test_monitoring.py:
```diff
@@ -81,6 +97,31 @@ def _check_object_exists():
+    volume_name = "test-volume-{}".format(random_salt)
+    client.create_from_yaml(DEFAULT_VOLUME.format(name=volume_name))
+
+    def _check_volume_ready():
+        volume = _get_volume_and_check_status(client, volume_name, 'Available')
+        assert volume['status'].get('deviceName') is not None, \
+            'Volume status.deviceName has not been reconciled'
+        return volume
+
+    yield utils.retry(
+        _check_volume_ready, times=30, wait=2,
+        name='waiting for Volume {} to be ready'.format(volume_name)
+    )
```
Can't reuse `check_volume_status` for that one?

gdemonet replied:
They are slightly different actually:
- this one checks if there's a device name - so it's a bit more than just checking the status phase
- `check_volume_status` is used in `test_volume` scenarii to manipulate the `context` fixture - we don't need it here

slaperche-scality replied:
Ok fair enough.

gdemonet committed:
tests: Factorize volume status checking

We extract common logic used to wait for a Volume to reach a given state
and add it to the VolumeClient utility.
We also extract the "test_volume" Given fixture, so that it uses this
new helper, and is now accessible to other test modules. We do not use
it in the Volume tests however, as these rely on finer control (in
particular, they store the resulting object in a `context` fixture).

bert-e commented:
# History mismatch



Merge commit #002feb7b315ba4f849464df8012f0d69a4a78732 on the integration branch
`w/2.6/improvement/2631-collect-metrics-from-partitions` is merging a branch which is neither the current
branch `improvement/2631-collect-metrics-from-partitions` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/2631-collect-metrics-from-partitions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/improvement/2631-collect-metrics-from-partitions` with contents from `improvement/2631-collect-metrics-from-partitions`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/improvement/2631-collect-metrics-from-partitions origin/development/2.6
 $ git merge origin/improvement/2631-collect-metrics-from-partitions
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/improvement/2631-collect-metrics-from-partitions
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #e9a8fd305cf90889f3bccab39420be6b89eb162f on the integration branch
`w/2.6/improvement/2631-collect-metrics-from-partitions` is merging a branch which is neither the current
branch `improvement/2631-collect-metrics-from-partitions` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/2631-collect-metrics-from-partitions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/improvement/2631-collect-metrics-from-partitions` with contents from `improvement/2631-collect-metrics-from-partitions`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/improvement/2631-collect-metrics-from-partitions origin/development/2.6
 $ git merge origin/improvement/2631-collect-metrics-from-partitions
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/improvement/2631-collect-metrics-from-partitions
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

gdemonet commented:
@slaperche-scality I had to rebase because of a bug in my test code. FYI, I was trying to resolve the real node name from what was defined in `spec.nodeName` of the Volume, but unlike what we need to do at creation time (i.e. replacing the "bootstrap" name in SSH config with the real node name we use), the Volume was already created hence this `nodeName` was already patched. So I could just use it as-is:

```python
    control_plane_ip = utils.get_grain(
        test_volume['spec']['nodeName'],
        'metalk8s:control_plane_ip'
    )
```

gdemonet commented:
> ```python
>     control_plane_ip = utils.get_grain(
>         test_volume['spec']['nodeName'],
>         'metalk8s:control_plane_ip'
>     )
> ```

Of course, this doesn't work either, because `utils.get_grain` expects a `host` object from testinfra... Need to think this through again.

gdemonet committed:
tests: Return from `run_salt_command` utility

This method was not returning its output, which we add here.
We also fix the assertion error message, which was not printing the
captured stderr (which is useful, when a command fails).

bert-e commented:
# History mismatch



Merge commit #6356f3717dd9c6201b4d3886707e1a38677fbf60 on the integration branch
`w/2.6/improvement/2631-collect-metrics-from-partitions` is merging a branch which is neither the current
branch `improvement/2631-collect-metrics-from-partitions` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/2631-collect-metrics-from-partitions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
```python
    # Retrieve control-plane IP of another Node through Salt master, not
    # through testinfra, because the actual Node name may not match our
    # SSH config file
    node_name = test_volume['spec']['nodeName']
    command = [
        'salt',
        '--out json',
        node_name,
        'grains.get',
        'metalk8s:control_plane_ip'
    ]
    result = utils.run_salt_command(host, command, ssh_config)
    control_plane_ip = json.loads(result.stdout)[node_name]
```

This one should work :crossed_fingers:

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/improvement/2631-collect-metrics-from-partitions` with contents from `improvement/2631-collect-metrics-from-partitions`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/improvement/2631-collect-metrics-from-partitions origin/development/2.6
 $ git merge origin/improvement/2631-collect-metrics-from-partitions
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/improvement/2631-collect-metrics-from-partitions
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

gdemonet commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/169dbca23701966626d1d4db729e0cc6c6628366) did not succeed in branch w/2.6/improvement/2631-collect-metrics-from-partitions.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye gdemonet.
